### PR TITLE
MAINT: Fix deprecated functions in scalarapi.c

### PR DESCRIPTION
**Pull request by charris**
For discussion.

Use of PyUnicode_DATA is deprecated in Python-3.9. The fix proposed here is to treat subclasses of NPY_UNICODE the same as NPY_UNICODE. That is how it was done before, and I really don't see what else we can do.

### CODE REVIEW

File: numpy/core/src/multiarray/scalarapi.c [hunk 1]
```diff
@@ -141,7 +141,18 @@ scalar_value(PyObject *scalar, PyArray_Descr *descr)
...
+            /* lazy initialization, to reduce the memory used by string scalars */
+            if (PyArrayScalar_VAL(scalar, Unicode) == NULL) {
+                Py_UCS4 *raw_data = PyUnicode_AsUCS4Copy(scalar);
+                if (raw_data == NULL) {
+                    return NULL;
+                }
+                PyArrayScalar_VAL(scalar, Unicode) = raw_data;
+                return (void *)raw_data;
+            }
+            return PyArrayScalar_VAL(scalar, Unicode);
```
Comment (mattip) [inline numpy/core/src/multiarray/scalarapi.c:155]: This is the same code as used above (line 74 or so) so the stanza itself is OK. The only other thing I could think of is to refactor `scalar_value()` to move the function itself onto the descriptor as a `__getscalar__` slot in the dtype refactoring.

**Comment by mattip** [comment]
LGTM. Will wait a bit if @seberg wants to comment.

File: numpy/core/src/multiarray/scalarapi.c (hunk 1 above)
Comment (seberg) [reply numpy/core/src/multiarray/scalarapi.c:155]: Yeah, we could do that and probably even should. I do not like `scalar_value()` as a requirement for scalars in general though.  And, I do not like flexible dtypes being special here, so would rather if we just return a `void *`.  In either case, Chuck probably wants to backport this, so we should put this in as is.

`scalar_value` (or something similar) is a neat idea.  But ultimately, it is mainly a slightly faster access to the value compared to `setitem` (i.e. one less copy).  The main problem for me is, that I do not really like it as a requirement on scalars (it technically is not even now), even if most scalars should be able to provide it easily.

**Comment by seberg** [comment]
Lets put this in, thanks.  I am not opposed to adding such a slot (with the possibility of it being NULL, and probably it being NULL for anything except builtin types for now).  The problem is, assessing possible project relying on it, but since user-dtypes are super rare right now, we can probably just get away with it and live with hard crashes in one or two places...

**Comment by charris** [comment]
I'd be happy to see `scalar_value` go if it doesn't serve any purpose.  For unicode the `PyUnicode_DATA` returned a pointer to void and could point to four different types. Not really that useful these days, in Python2 one knew what it pointed to depending on a wide or narrow build. Deprecating it was the right way to go. But this at least gets rid of the compiler warning. The one remaining warning is gone using cython master, the released version don't seem to be there yet, although perhaps we just need to upgrade. @mattip should we upgrade cython for testing?

**Comment by mattip** [comment]
I think we need to wait for Cyton 0.29.21 to be released, at least according to the changelog http://docs.cython.org/en/latest/src/changes.html#id13

**Pull request merged**